src/events.ts

```diff
@@ -0,0 +1,86 @@
+export type TurboClickEvent = CustomEvent<{ url: string; originalEvent: MouseEvent }>
+export type TurboFrameLoadEvent = CustomEvent
+export type TurboBeforeFrameRenderEvent = CustomEvent<{ newFrame: FrameElement } & FrameViewRenderOptions>
+export type TurboFrameRenderEvent = CustomEvent<{ fetchResponse: FetchResponse }>
+export type TurboLoadEvent = CustomEvent<{ url: string; timing: TimingData }>
+export type TurboRenderEvent = CustomEvent
+export type TurboReloadEvent = CustomEvent
+export type TurboVisitEvent = CustomEvent<{ url: string; action: Action }>
+
+export type TurboBeforeStreamRenderEvent = CustomEvent<{ newStream: StreamElement; render: Render }>
+
+export type TurboSubmitStartEvent = CustomEvent<{ formSubmission: FormSubmission }>
+export type TurboSubmitEndEvent = CustomEvent<
+  { formSubmission: FormSubmission } & { [K in keyof FormSubmissionResult]?: FormSubmissionResult[K] }
+>
+
+export type TurboFrameMissingEvent = CustomEvent<{ response: Response; visit: VisitFallback }>
+
+export type TurboBeforeFetchRequestEvent = CustomEvent<{
+  fetchOptions: RequestInit
+  url: URL
+  resume: (value: any) => void
+}>
+
+export type TurboBeforeFetchResponseEvent = CustomEvent<{
+  fetchResponse: FetchResponse
+}>
+
+export type TurboFetchRequestErrorEvent = CustomEvent<{
+  request: FetchRequest
+  error: Error
+}>
+
+interface TurboGlobalEventMap {
+  "turbo:before-cache": TurboBeforeCacheEvent
+  "turbo:before-fetch-request": TurboBeforeFetchRequestEvent
+  "turbo:before-fetch-response": TurboBeforeFetchResponseEvent
+  "turbo:before-frame-render": TurboBeforeFrameRenderEvent
+  "turbo:before-stream-render": TurboBeforeStreamRenderEvent
+  "turbo:before-render": TurboBeforeRenderEvent
+  "turbo:before-visit": TurboBeforeVisitEvent
+  "turbo:click": TurboClickEvent
+  "turbo:fetch-request-error": TurboFetchRequestErrorEvent
+  "turbo:frame-load": TurboFrameLoadEvent
+  "turbo:frame-render": TurboFrameRenderEvent
+  "turbo:frame-missing": TurboFrameMissingEvent
+  "turbo:load": TurboLoadEvent
+  "turbo:render": TurboRenderEvent
+  "turbo:reload": TurboReloadEvent
+  "turbo:submit-start": TurboSubmitStartEvent
+  "turbo:submit-end": TurboSubmitEndEvent
+  "turbo:visit": TurboVisitEvent
+}
+
+interface TurboElementEventMap {
+  "turbo:before-frame-render": TurboBeforeFrameRenderEvent
+  "turbo:before-fetch-request": TurboBeforeFetchRequestEvent
+  "turbo:before-fetch-response": TurboBeforeFetchResponseEvent
+  "turbo:fetch-request-error": TurboFetchRequestErrorEvent
+  "turbo:frame-load": TurboFrameLoadEvent
+  "turbo:frame-render": TurboFrameRenderEvent
+  "turbo:frame-missing": TurboFrameMissingEvent
+  "turbo:submit-start": TurboSubmitStartEvent
+  "turbo:submit-end": TurboSubmitEndEvent
+}
+
+declare global {
+  interface DocumentEventMap extends TurboGlobalEventMap {}
+  interface WindowEventMap extends TurboGlobalEventMap {}
+  interface ElementEventMap extends TurboElementEventMap {}
```

Comment:
Should the `TurboElementEventMap` extend from [HTMLElementEventMap](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.htmlelementeventmap.html) instead of [ElementEventMap](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.elementeventmap.html)?

Reply:
Similarly, could the [DocumentEventMap](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.documenteventmap.html) and [WindowEventMap](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.windoweventmap.html) declarations be combined into a single extension of the [GlobalEventHandlersEventMap](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.globaleventhandlerseventmap.html) interface?

Reply:
I picked `ElementEventMap` on purpose since we've been using `Element` over `HTMLElement` almost everywhere, but if you think we just just limit it to `HTMLElement` let me know.

OH, I didn't know about `GlobalEventHandlersEventMap`, maybe it would make sense to use that then, thanks for the hint!